pkg/gofr/datasource/opentsdb/response.go

```diff
@@ -0,0 +1,439 @@
+
+			message = fmt.Sprint(errUnmarshaling)
+
+			return errUnmarshaling
+		}
+	} else {
+		// Use the custom parser if available.
+		if err := parser(jsonBytes); err != nil {
+			message = fmt.Sprintf("failed to parse response body through custom parser %s %s: %v", method, url, err)
+			return err
+		}
+	}
+
+	status = StatusSuccess
+	message = fmt.Sprintf("%s request sent at : %s", method, url)
+
+	return nil
+}
+
+func (c *Client) version(ctx context.Context, verResp *VersionResponse) error {
+	span := c.addTrace(ctx, "Version")
+
+	status := StatusFailed
+
+	var message string
+
+	defer sendOperationStats(c.logger, time.Now(), "Version", &status, &message, span)
+
+	verEndpoint := fmt.Sprintf("%s%s", c.endpoint, VersionPath)
+	verResp.logger = c.logger
+	verResp.tracer = c.tracer
+	verResp.ctx = ctx
+
+	if err := c.sendRequest(ctx, http.MethodGet, verEndpoint, "", verResp); err != nil {
+		message = fmt.Sprintf("error while processing request at URL %s: %s", verEndpoint, err)
+		return err
+	}
+
+	status = StatusSuccess
+	message = "version response retrieved successfully."
+
+	return nil
+}
+
+// isValidOperateMethod checks if the provided HTTP method is valid for
+// operations such as POST, PUT, or DELETE.
+func (c *Client) isValidOperateMethod(ctx context.Context, method string) bool {
+	span := c.addTrace(ctx, "isValidOperateMethod")
+
+	status := StatusSuccess
+
+	var message string
+
+	defer sendOperationStats(c.logger, time.Now(), "isValidOperateMethod", &status, &message, span)
+
+	method = strings.TrimSpace(strings.ToUpper(method))
+	if method == "" {
+		return false
+	}
+
+	validMethods := []string{http.MethodPost, http.MethodPut, http.MethodDelete, http.MethodPut}
+	for _, validMethod := range validMethods {
+		if method == validMethod {
+			return true
+		}
+	}
+
+	return false
+}
+
+func customParserHelper(ctx context.Context, resp genericResponse, operation string, logger Logger,
+	unmarshalFunc func([]byte) error) func([]byte) error {
+	span := resp.addTrace(ctx, operation)
+
+	status := StatusFailed
+
+	var message string
+
+	defer sendOperationStats(logger, time.Now(), operation, &status, &message, span)
+
+	return func(result []byte) error {
+		err := unmarshalFunc(result)
+		if err != nil {
+			message = fmt.Sprintf("unmarshal %s error: %s", operation, err)
+			logger.Errorf(message)
+
+			return err
+		}
+
+		status = StatusSuccess
+		message = fmt.Sprintf("%s custom parsing was successful.", operation)
+
+		return nil
+	}
+}
+
+func queryParserHelper(ctx context.Context, logger Logger, obj genericResponse, methodName string) func(respCnt []byte) error {
+	return customParserHelper(ctx, obj, methodName, logger, func(resp []byte) error {
+		originRespStr := string(resp)
+
+		var respStr string
+
+		if len(resp) != 0 && resp[0] == '[' && resp[len(resp)-1] == ']' {
+			respStr = fmt.Sprintf(`{"queryRespCnts":%s}`, originRespStr)
+		} else {
+			respStr = originRespStr
+		}
+
+		return json.Unmarshal([]byte(respStr), obj)
```

Comment:
I dislike how you do, but if you want to keep this, you can do this at least

```suggestion
		if strings.HasPrefix(resp, "[") && strings.HasSuffix(resp, "]") {
			respStr = fmt.Sprintf(`{"queryRespCnts":%s}`, originRespStr)
		} else {
			respStr = originRespStr
		}

		return json.Unmarshal([]byte(respStr), obj)
```

I would have done something like this, instead of parsing the data like this


```suggestion

		var cnts []any
		err := json.Unmarshal([]byte(originRespStr), &cnts)
		if err != nil {
			return json.Unmarshal([]byte(originRespStr), obj)
		}
		obj. = cnts
		return nil
```

Reply:
Sure changed it to this as per your suggestion:

```go
var respStr string

		if strings.HasPrefix(string(resp), "[") && strings.HasSuffix(string(resp), "]") {
			respStr = fmt.Sprintf(`{"queryRespCnts":%s}`, originRespStr)
		} else {
			respStr = originRespStr
		}

		return json.Unmarshal([]byte(respStr), obj)
```